docs/docs/design/bitfield-signing.md

```diff
@@ -0,0 +1,90 @@
+# Bitfield Signing Subsystem
+
+[The bitfield signing subsystem](https://paritytech.github.io/polkadot-sdk/book/node/availability/bitfield-signing.html)
+is responsible for producing signed availability bitfields and passing them to
+[the bitfield distribution subsystem](./bitfield-distribution.md). It only does that when running on a validator.
+
+## Subsystem Structure
+
+The implementation must conform to the `Subsystem` interface defined in the `parachaintypes` package. It should live in
+a package named `bitfieldsigning` under `dot/parachain/bitfield-signing`.
+
+### Messages Received
+
+The subsystem must be registered with the overseer. It only handles `parachaintypes.ActiveLeavesUpdateSignal`. The
+overseer must be modified to forward this message to the subsystem.
+
+### Messages Sent
+
+1. [`DistributeBitfield`](https://github.com/paritytech/polkadot-sdk/blob/1e3b8e1639c1cf784eabf0a9afcab1f3987e0ca4/polkadot/node/subsystem-types/src/messages.rs#L522)
+
+```go
+package parachaintypes
+
+import "github.com/ChainSafe/gossamer/lib/common"
+
+type DistributeBitfield struct {
+	RelayParent common.Hash
+	Bitfield    UncheckedSignedAvailabilityBitfield
+}
+```
+
+As mentioned in [bitfield distribution](./bitfield-distribution.md), it makes sense to duplicate the type
+`UncheckedSignedAvailabilityBitfield` as `CheckedSignedAvailabilityBitfield` or just `SignedAvailabilityBitfield` and
+use it in `DistributeBitfield`.
+
+2. [`AvailabilityStore::QueryChunkAvailability(CandidateHash, ValidatorIndex, response_channel)`](https://github.com/paritytech/polkadot-sdk/blob/1e3b8e1639c1cf784eabf0a9afcab1f3987e0ca4/polkadot/node/subsystem-types/src/messages.rs#L556)
+
+This type is already defined as `availabilitystore.QueryChunkAvailability` in
+`dot/parachain/availability-store/messages.go`.
+
+## Subsystem State
+
+The subsystem needs access to the parachain session keys for signing bitfields. The appropriate `Keystore` should be
+passed to the subsystem constructor as a dependency to facilitate testing. Since each new active leave results in a
+goroutine being spawned, the subsystem needs to maintain state to receive data from them and to cancel them when
+appropriate.
+
+```go
+package bitfieldsigning
+
+import (
+	"context"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/keystore"
+)
+
+type signingTask struct {
+	ctx      context.Context
+	response <-chan parachaintypes.UncheckedSignedAvailabilityBitfield
+}
+
+type BitfieldSigning struct {
+	keystore keystore.Keystore
```

Comment:
is this field shared among the spawned goroutines? do we need to make it safe against race conditions?

Reply:
Good point! I think it is possible to avoid this and added a paragraph describing how. Since signing should not mutate the private key, I think this would work.